File: CSharp/Shapes/Plane.cs
using System;

namespace CSharp
{
    public class Plane : RTObject
    {
        private BoundingBox _bounds;

        public Plane()
        {
            _bounds = new BoundingBox(
                                        new Point(double.NegativeInfinity, 0, double.NegativeInfinity),
                                        new Point(double.PositiveInfinity, 0, double.PositiveInfinity)
                                      );
        }

        public  override BoundingBox BoundsOf => _bounds;

        protected override Intersections LocalIntersect(Ray ray)
        {
            var i = new Intersections();

            if (Math.Abs(ray.Direction.y) < MathHelper.Epsilon)
            {
                return i;
            }

            var t = -ray.Origin.y / ray.Direction.y;
            i.Add(new Intersection(t, this));

            return i;
        }

        protected override Vector LocalNormal(Point p, Intersection hit)
        {
            return Vector.VectorY;
        }

        public override bool Equals(object obj)
        {
            var other = obj as Plane;
            return base.Equals(obj);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine("Plane", base.GetHashCode());
        }

        public override string ToString()
        {
            return $"Plane ({ID}): {Material} {Transform}";
        }
    }
}

File: Java/src/main/java/Shapes/Plane.java
package RTC.Shapes;

import RTC.Math.*;
import RTC.Shapes.Shape;
import RTC.Material.*;
import java.util.Objects;

public class Plane extends Shape {
    public Plane() {
        super();
    }

    public Vector localNormalAt(Point localPoint) {
        return Vector.VectorY;
    }

    public Intersections localIntersect(Ray r) {
        Intersections result = new Intersections();

        if (Math.abs(r.getDirection().getY()) < Helper.EPSILON) {
            return result;
        }

        double t = -r.getOrigin().getY() / r.getDirection().getY();
        result.add(new Intersection(t, this));

        return result;
    }

    @Override
    public boolean equals(Object o)
    {
        if (o == this) {
            return true;
        }

        if ((o == null) || !(o instanceof Plane)) {
            return false;
        }

        Plane r = (Plane) o;
        return true;
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return String.format("Plane Id: %d", id);
    }
}
